language/move-compiler/src/to_bytecode/optimize/mod.rs

```diff
@@ -0,0 +1,39 @@
+// Copyright (c) The Move Contributors
+// SPDX-License-Identifier: Apache-2.0
+
+use crate::parser::ast::FunctionName;
+use move_ir_types::ast::{self as IR};
+use std::collections::{BTreeSet, HashMap};
+
+mod remove_fallthrough_jumps;
+mod remove_unnecessary_store;
+mod remove_unused_locals;
+
+pub(crate) fn code(
+    _f: &FunctionName,
+    loop_heads: &BTreeSet<IR::BlockLabel_>,
```

Comment:
Is there a risk that this could become out-of-date with the blocks?  It seems generally possible given the interface here, so we should document this function with the kind of optimisations it does.

Reply:
In general yes, but I don't think it is currently. The inlinining of blocks only happens if there is a single jump target. 
If the loop is the beginning of the function (the head is the first block), it won't get inlined.
Otherwise it will have two jumps, the path to the loop and the backedge

Reply:
Is it possible for a non-loop block to get inlined and then shift all the indices around?  I know at least somewhere in the compiler block labels are just code offsets, so I'm imagining a scenario like:


```
B1: ...
    Branch B5
B2: ...
    BrFalse B4
B3: ...
    Branch B2
B4: ...
    Ret
B5: ...
    Branch B2
```

`B2` is a loop head, and `B5` can be merged with `B1` to shift its code offset.  I know that the way the compile generates code makes it at least unlikely that blocks could be laid out this way, but I couldn't convince myself (especially in the presence of optimisations) that it was impossible.